### PR TITLE
rewrite disko in C++ (WIP)

### PR DESCRIPTION
**Pull request by Sk7Str1p3**
@iFreilicht as i promised, my current code. for now, only basic functions implemented. it feels bit shitty for me, waiting for response and advice , thx.

### CODE REVIEW

File: src/disko/cli.cpp [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+      exit(1);
+    }
+  }
+  return disko_Mode;
+  return flake_Path;
+  return mount_Path;
+  return nix_Args;
+  return dry_Run;
+  return omit_Check;
+  return verbose;
```
Comment (iFreilicht) [inline src/disko/cli.cpp:134]: Are multi-returns like this in C++ even an actual feature? I would  expect the first return to immediately end the function.

**Comment by iFreilicht** [comment]
I think you might have misunderstood what "rewrite" means in the context of #789. A rewrite of just the CLI entrypoint in another language will solve nothing. We need to re-implement all the `_create` options in all the `lib/types/*.nix` files. These contain the bash code that's so hard to maintain. In addition, I want to use this opportunity to change to a different execution strategy that allows a user to verify which steps are executed and why before running them.

If you want to make a case for C++ being a good language for this rewrite, you need to show a PoC of that. This is a high ask, I know, but I've been working through multiple iterations of the Python rewrite and haven't even reached that point yet; I myself have not yet shown that a rewrite in Python is even feasible.

File: src/disko/cli.cpp (hunk 1 above)
Comment (Sk7Str1p3) [reply src/disko/cli.cpp]: fixed, now function adds variables into array and returns it.

**Comment by Sk7Str1p3** [comment]
> I think you might have misunderstood what "rewrite" means in the context of #789. A rewrite of just the CLI entrypoint in another language will solve nothing. We need to re-implement all the `_create` options in all the `lib/types/*.nix` files. 

No, i understood it correctly. CLI entrypoint is just the beginning of this, I'm going to reimplement those functions and make other minor changes.

**Comment by iFreilicht** [comment]
Okay, good. But in that case I'll unsubscribe from this PR and would ask you to mention me again if anything interesting comes up. I don't want to get pinged every time you push a commit.

**Comment by Mic92** [comment]
If I am honest, I don't feel comfortable maintaining disko if it was written in C++. I maintain a few C++ project and I run into way too many footguns and we would have the issue that we have to compile inside an installer when used in nixos-anywhere beyond just the entrypoint CLI.
If you want to go this route, I would suggest forking the project. We can link to it from here.

**Comment by Sk7Str1p3** [comment]
Wouldn't we have to compile inside an installer in any way? I don't think this is C++ specific problem. Also (at least for now) maintaining C++ code feels not so bad and hard. 

However, if you insist, I can rewrite this in other language. But which one?

**Comment by Enzime** [comment]
My preference would be Rust over C++, however the lower bar to entry in Python is also nice

If we just passed a JSON blob to a compiled program then you wouldn’t need to build it locally if it was cached by Hydra/packaged in Nixpkgs

**Comment by Mic92** [comment]
The rust toolchain is also too big to be downloaded in an in-memory installer. Further details in https://github.com/nix-community/disko/issues/789

**Comment by Mic92** [comment]
I think, I prefer the python approach that @iFreilicht is following. pythonMinimal is still way smaller to upload than a whole gcc toolchain.
If you want to create and maintain your own an alternative disko implementation, we can link it from our README.

**Comment by Sk7Str1p3** [comment]
@Mic92 python **requires** toolchain to be downloaded while gcc toolchain is **included** in NixOS live iso by default.
 
Also, as I said previously, we can use binary cache with C++, but we **can't** go that way with python. Using C++ seems more logical for this.

**Comment by Sk7Str1p3** [comment]
Then why you were offering rewrite Disko in Rust? It has literally same syntax with C++

**Pull request closed without merging**